### PR TITLE
Cleanup `rmeta::MacroDef`

### LINKED ISSUE #49511: macro_rules macros should not be serialized through strings in crate metadata.
`macro_rules` should use `TokenStream`s instead of being pretty-printed into a string and later re-parsed. This is required to get correct/precise hygiene information and original source spans.

For example, right now macros from `libstd` get `<std macros>` in their spans instead of the original source, and we hide printing any of that source because it looks much worse than the original, and those spans also don't capture any fine-grained information from macro-defining macros.

IIRC @ibabushkin started working on this (see https://github.com/rust-lang/rust/pull/43847) but I couldn't find a tracking issue.
cc @alexcrichton @jseyfried @Manishearth

### PR DESCRIPTION
**Pull request by Centril**
Avoid using rountrip parsing in the encoder and in `fn load_macro_untracked`.

The main reason I was interested in this was to remove `rustc_parse` as a dependency of `rustc_metadata` but it seems like this had other benefits as well.

Fixes #49511.

r? @eddyb
cc @matthewjasper @estebank @petrochenkov

### CODE REVIEW

File: src/test/ui/borrowck/issue-64453.stderr
Comment (Mark-Simulacrum) [inline src/test/ui/borrowck/issue-64453.stderr]: This... seems not great. I don't think it helps people to see this, and in fact is probably more confusing for the average person?

File: src/test/ui/match/match-arm-resolving-to-never.stderr
Comment (Mark-Simulacrum) [inline src/test/ui/match/match-arm-resolving-to-never.stderr]: e.g. this is exposing `std::rt` as a module and this function which are intentionally private and unstable...

File: src/librustc_metadata/rmeta/decoder.rs
Comment (Mark-Simulacrum) [inline src/librustc_metadata/rmeta/decoder.rs]: Out of interest -- what caused this to be necessary? Do you know? I can't find a diff in this PR which added the Session parameter to some function...

File: src/test/ui/borrowck/issue-64453.stderr
Comment (Centril) [reply src/test/ui/borrowck/issue-64453.stderr]: Not sure -- ostensibly it could be more helpful because it highlights the reason behind the error...

**Review by Mark-Simulacrum — COMMENTED** [review]
Overall though I think this is reasonable, I guess :)

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (Centril) [inline src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: ...for example I think this is clearer.

File: src/test/ui/match/match-arm-resolving-to-never.stderr
Comment (Centril) [reply src/test/ui/match/match-arm-resolving-to-never.stderr]: Agreed on this one for sure tho. :)

File: src/test/ui/borrowck/issue-64453.stderr
Comment (Mark-Simulacrum) [reply src/test/ui/borrowck/issue-64453.stderr]: I don't get any more information to be honest from this underlying function/macro and it's sort of quite "raw" and such

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (Mark-Simulacrum) [reply src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: Kind of. On the other hand, in the general case of cross-crate macros (which I believe this is), this span doesn't actually help you any since you can't fix it without an upstream PR...

File: src/librustc_metadata/rmeta/decoder.rs
Comment (Centril) [reply src/librustc_metadata/rmeta/decoder.rs]: I got an ICE when I just had `.decode(self)`. It arose from `impl<'a, 'tcx> SpecializedDecoder<Span> for DecodeContext<'a, 'tcx> {`. You can grep for `bug!("Cannot decode Span without Session.")`. @eddyb suggested I change this. :)

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (Centril) [reply src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: I guess it depends on whether you own the upstream crate or whether you would feel comfortable making said PR. If you would, then it could help?

This is not intentional from my end tho; just a side-effect of removing the hack which I was pointed to :)

File: src/librustc_metadata/rmeta/decoder.rs
Comment (eddyb) [reply src/librustc_metadata/rmeta/decoder.rs]: The `decode` method is generic, and in its most basic form doesn't require a `&Session` - perhaps we can make that mandatory?

But the change here is in line with other methods in this file, you'll see what I mean if you look around for a bit.

File: src/test/ui/thread-local-not-in-prelude.rs
Comment (eddyb) [inline src/test/ui/thread-local-not-in-prelude.rs]: Wow, so the unused variable lint wasn't triggering before?

Did you only have to make this change before you fixed the lint logic, though?

Comment (Centril) [reply src/test/ui/thread-local-not-in-prelude.rs]: Oops; forgot to revert this -- should be done now when I push again (after figuring out the other problem...).

**Comment by Centril** [comment]
All the tests passed (before I adjusted the comment), so this should be ready for review now. :)

**Comment by eddyb** [comment]
I looked around and I found some parts of the code that can be cleaned up.

These can all be removed:
https://github.com/rust-lang/rust/blob/d63b24ffcc48f44ef09e0369e6516d6f2dec3520/src/libsyntax_pos/lib.rs#L69-L70
https://github.com/rust-lang/rust/blob/d63b24ffcc48f44ef09e0369e6516d6f2dec3520/src/libsyntax_pos/lib.rs#L93
https://github.com/rust-lang/rust/blob/d63b24ffcc48f44ef09e0369e6516d6f2dec3520/src/libsyntax_pos/lib.rs#L131-L145

And this should probably be refactored, to take into account that `is_macros()` is never `true`:
https://github.com/rust-lang/rust/blob/d63b24ffcc48f44ef09e0369e6516d6f2dec3520/src/librustc_errors/emitter.rs#L270-L396

**Comment by petrochenkov** [comment]
Do not forget to run perf on this.
Stringification was a cheat, but it could be a performance-friendly cheat.

**Comment by Centril** [comment]
Let's try a perf run now (and maybe one later after doing some more changes according to https://github.com/rust-lang/rust/pull/66364#issuecomment-553934604):

@bors try @rust-timer queue

**Comment by Centril** [comment]
Looks like perf has clearly improved a small bit. @bors rollup=never

**Comment by petrochenkov** [comment]
Nice.

**Comment by Dylan-DPC** [comment]
@Centril is this waiting on anything from the author? Or did you forget to r+ it?

**Comment by Centril** [comment]
@Dylan-DPC I'm the author. 😄 It's waiting on me to address https://github.com/rust-lang/rust/pull/66364#issuecomment-553934604.

**Comment by Dylan-DPC** [comment]
Ah cool (realised it after posting and updated my comment )

**Comment by Alexendoo** [comment]
Ping from triage, any updates? @Centril

**Comment by Centril** [comment]
@eddyb I made some changes in https://github.com/rust-lang/rust/pull/66364/commits/8ee0a752ae5f7493ccca718f5bccbe70b3d4b3ca. Unfortunately, it seems to cause a problem with rustfix in `ui/suggestions/vec-macro-in-pattern.rs`. Would appreciate some help here... I don't feel like I know what I'm doing. If you can do things faster, please consider taking over the branch. :)

File: src/test/ui/suggestions/format-borrow.stderr
Comment (eddyb) [inline src/test/ui/suggestions/format-borrow.stderr]: Hmm can we avoid this in suggestions? It's definitely doing the wrong thing, looks like it's mixing source code from two spans?

File: src/test/ui/suggestions/dont-suggest-try_into-in-macros.stderr
Comment (eddyb) [inline src/test/ui/suggestions/dont-suggest-try_into-in-macros.stderr]: Another suggestion printing macro code.

File: src/test/ui/suggestions/dont-suggest-deref-inside-macro-issue-58298.stderr
Comment (eddyb) [inline src/test/ui/suggestions/dont-suggest-deref-inside-macro-issue-58298.stderr]: Based on the name of the test, this is exactly that's undesired.

File: src/test/ui/span/coerce-suggestions.stderr
Comment (eddyb) [inline src/test/ui/span/coerce-suggestions.stderr]: Same here.

File: src/test/ui/proc-macro/three-equals.stderr
Comment (eddyb) [inline src/test/ui/proc-macro/three-equals.stderr]: @petrochenkov This is an improvement, right?

File: src/test/ui/proc-macro/mixed-site-span.stderr
Comment (eddyb) [inline src/test/ui/proc-macro/mixed-site-span.stderr]: Not sure why "in this macro invocation" messages are going away.

File: src/test/ui/privacy/associated-item-privacy-type-binding.rs
Comment (eddyb) [inline src/test/ui/privacy/associated-item-privacy-type-binding.rs]: Accidentally committed to this PR? Or does it have something to do with suggestions?

File: src/test/ui/methods/method-on-ambiguous-numeric-type.stderr
Comment (eddyb) [inline src/test/ui/methods/method-on-ambiguous-numeric-type.stderr]: More suggestions inside macros?

File: src/test/ui/diverging-tuple-parts-39485.stderr
Comment (eddyb) [inline src/test/ui/diverging-tuple-parts-39485.stderr]: Ohh, I think I see what's happening, it's looking at the macro span now instead of the invocation span.

Maybe it's using some other heuristic we didn't think of yet?

File: src/test/ui/privacy/associated-item-privacy-type-binding.rs
Comment (estebank) [reply src/test/ui/privacy/associated-item-privacy-type-binding.rs]: Probably to avoid warning output unrelated to what's being tested.

File: src/test/ui/proc-macro/span-api-tests.rs
Comment (eddyb) [inline src/test/ui/proc-macro/span-api-tests.rs]: Actually, now that I understand what this was doing (checking `proc_macro::SourceFile::is_real` returning `false`), maybe `say_hello_extern` should be replaced with a proc macro? cc @petrochenkov

**Comment by eddyb** [comment]
Status: turns out `fix_multispans_in_std_macros` is hacks on top of hacks, and it contains the backtrace printing logic as a "side" thing, but also `-Z external-macro-backtrace` is misnamed because it acts on all backtraces, it's just the message to pass it that depends on local vs external.

I think I'll open a separate PR to fix this (how? have to pick between always showing backtrace for local macros, or renaming the flag - I'm leaning towards the latter).

File: src/test/ui/proc-macro/mixed-site-span.stderr
Comment (eddyb) [reply src/test/ui/proc-macro/mixed-site-span.stderr]: Finally figured it out, "in this macro invocation" should be skipped when it's on the same span as the message itself.

Why can that happen? Well, it's *specifically* because we replace spans inside external macros (and after this PR inside any cross-crate-imported files) with the macro call site, so that we don't show implementation details.

If the span would be identical, the message is redundant since it doesn't help anything. But the logic is confusing because this is not explicit anywhere, it just happens to check the same thing in multiple places!

Hopefully I can clean this up and comment it a bit.

**Comment by petrochenkov** [comment]
This PR may fix https://github.com/rust-lang/rust/issues/49511 (depending on interpretation).

File: src/librustc/lint.rs
Comment (eddyb) [inline src/librustc/lint.rs]: Does anything change if you remove the dummy case? Dummy spans might be irrelevant nowadays.

File: src/librustc_errors/emitter.rs
Comment (eddyb) [inline src/librustc_errors/emitter.rs]: At the very least, `<*macros>` in comments should be replaced with "external macros" or so.

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (eddyb) [inline src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: Maybe the test should be changed to have `pub fn async` on one line, so the diagnostic output makes more sense? Or am I missing something?

File: src/test/ui/editions/edition-keywords-2018-2018-expansion.stderr
Comment (eddyb) [inline src/test/ui/editions/edition-keywords-2018-2018-expansion.stderr]: Same thing here.

File: src/test/ui/proc-macro/span-api-tests.rs
Comment (eddyb) [reply src/test/ui/proc-macro/span-api-tests.rs]: I think this may still be relevant.

File: src/test/ui/methods/method-on-ambiguous-numeric-type.stderr
Comment (eddyb) [reply src/test/ui/methods/method-on-ambiguous-numeric-type.stderr]: This is actually pre-existing, seems consistent with the error just above in the same file, don't think this needs any changes.

File: src/librustc/lint.rs
Comment (eddyb) [reply src/librustc/lint.rs]: Or maybe we can leave it alone for now.

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (Centril) [reply src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: Formatting the aux testfile seems to have no effect.

File: src/librustc/lint.rs
Comment (Centril) [reply src/librustc/lint.rs]: Yeah I would prefer to do less. :)

File: src/librustc_errors/emitter.rs
Comment (Centril) [reply src/librustc_errors/emitter.rs]: Done.

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (eddyb) [reply src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: I don't understand, that shouldn't be possible.

File: src/test/ui/editions/auxiliary/edition-kw-macro-2015.rs
Comment (eddyb) [inline src/test/ui/editions/auxiliary/edition-kw-macro-2015.rs]: This is in the wrong direction. The error output looks like `async` is on its own line and I wanted `pub fn async` to not be split into multiple lines.

Is it possible the error reporting itself is doing something weird here, and printing `r#async` without it actually being a code snippet?

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (eddyb) [reply src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: Just saw your changes, `pub fn async` was already on one line, I guess I wasn't clear enough (I should've just checked).

**Comment by eddyb** [comment]
@petrochenkov I am trying to come up with a way for @Centril to rewrite `say_hello_extern` (from `src/test/ui/proc-macro/span-api-tests.rs`) as a proc macro that generates the `"Hello, World!"` literal with a `Span` that doesn't point to a real `FileName`.

I naively expected, judging from this, that the `FileName` would be `ProcMacroSourceCode`, for tokens parsed from a string (e.g. `"foo bar baz".parse::<TokenStream>()`): https://github.com/rust-lang/rust/blob/e6ec0d125eba4074122b187032474b4174fb9d31/src/librustc_expand/proc_macro_server.rs#L397-L404

But the last argument is named `override_span` in `parse_stream_from_source_str`, so I guess we try very hard to hide what the actual source was?

**Comment by Centril** [comment]
@petrochenkov Ping in case you missed https://github.com/rust-lang/rust/pull/66364#issuecomment-584317587.

(I'm not sure how to progress this PR on my end; if any of you two can take over I'd be grateful.)

**Comment by petrochenkov** [comment]
I indeed missed https://github.com/rust-lang/rust/pull/66364#issuecomment-584317587.

Looks like significant part of this PR can be landed without the primary change (if it's moved to a separate PR) - `#[allow]`s, `is_macros` -> `is_imported`, comments, `get_macro`, tests.

I'll get to the primary part this weekend, most probably.

File: src/test/ui/editions/auxiliary/edition-kw-macro-2015.rs
Comment (petrochenkov) [reply src/test/ui/editions/auxiliary/edition-kw-macro-2015.rs]: I didn't immediately figure out what eddyb meant in this commit, but it appears to be this diff:
```diff
-LL | () => (pub fn r#async () { })
+LL | r#async
```
Error reporting machinery no longer fetches the whole `r#async`'s line for displaying, only the `r#async` itself.

File: src/test/ui/proc-macro/span-api-tests.rs
Comment (petrochenkov) [reply src/test/ui/proc-macro/span-api-tests.rs]: It's quite possible that proc macros are no longer "fake source files" either, since they keep their spans in metadata.
I agree that we need to attempt making `say_hello_extern` a proc macro, and if `assert_fake_source_file` still fails, then remove the test case.

**Comment by petrochenkov** [comment]
Remaining unresolved comments:
https://github.com/rust-lang/rust/pull/66364#discussion_r357677628
https://github.com/rust-lang/rust/pull/66364#discussion_r377254283
https://github.com/rust-lang/rust/pull/66364#issuecomment-584317587

**Comment by petrochenkov** [comment]
@eddyb 
>But the last argument is named override_span in parse_stream_from_source_str, so I guess we try very hard to hide what the actual source was?

That's true, looks like `FileName` is never used if `override_span` is provided, perhaps they could be packed into some enum to show that they are never used at the same time.
Proc macros always provide an override span, so `ProcMacroSourceCode` is dead code.
Other `FileName` variants need to be audited as well, maybe they are also dead.

The variants that are not dead can be used for the test if spans from them can be tested in UI tests at all, otherwise the test case can be deleted.

File: src/test/ui/proc-macro/span-api-tests.rs
Comment (petrochenkov) [reply src/test/ui/proc-macro/span-api-tests.rs]: Answered in https://github.com/rust-lang/rust/pull/66364#issuecomment-590095860, spans from proc macros are always real.

File: src/test/ui/editions/auxiliary/edition-kw-macro-2015.rs
Comment (Centril) [reply src/test/ui/editions/auxiliary/edition-kw-macro-2015.rs]: So I tried:
```rust

        let rmeta::MacroDef { body, legacy } = data.get_macro(id.index, sess);
        let span = data.get_span(id.index, sess);

        let macro_full_name = data.def_path(id.index).to_string_friendly(|_| data.root.name);
        let source_name = FileName::Custom(macro_full_name);
        let source_content = rustc_ast_pretty::pprust::tts_to_string(body.clone());
        let source_file = sess.parse_sess.source_map().new_source_file(source_name, source_content);
        let span = span.with_lo(source_file.start_pos).with_hi(source_file.end_pos);
```
to see if the primary change was the cause, but it didn't result in reverting this case, so I'm thinking that's not the cause.

So at this point I don't really have a clue what could have caused this diff.

**Comment by eddyb** [comment]
@Centril You can remove https://github.com/rust-lang/rust/pull/66364/commits/0a114dcda56dd3db8f3cc4ea6dfe617e864fba7b, since it was based on my misunderstanding.

**Comment by Centril** [comment]
I dropped the last formatting commit and ran:
```rust
rustc +gamma-stage1 src/test/ui/editions/edition-keywords-2015-2018-expansion.rs --extern edition_kw_macro_2018=./build/x86_64-unknown-linux-gnu/test/ui/editions/edition-keywords-2015-2018-expansion/auxiliary/libedition_kw_macro_2018.so
```
which produced:
```rust
error: expected identifier, found keyword `async`
 --> src/test/ui/editions/edition-keywords-2015-2018-expansion.rs:8:5
  |
8 |     produces_async! {} //~ ERROR expected identifier, found keyword
  |     ^^^^^^^^^^^^^^^^^^ expected identifier, found keyword
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: you can escape reserved keywords to use them as identifiers
  |
7 | r#async
  |

error: aborting due to previous error
```
and then I reran with `-Z macro-backtrace` added and it produced:
```rust
error: expected identifier, found keyword `async`
 --> /home/centril/programming/rust/rust-gamma/src/test/ui/editions/auxiliary/edition-kw-macro-2018.rs:7:19
  |
6 | / macro_rules! produces_async {
7 | |     () => (pub fn async() {})
  | |                   ^^^^^ expected identifier, found keyword
8 | | }
  | |_- in this expansion of `produces_async!`
  | 
 ::: src/test/ui/editions/edition-keywords-2015-2018-expansion.rs:8:5
  |
8 |       produces_async! {} //~ ERROR expected identifier, found keyword
  |       ------------------ in this macro invocation
  |
help: you can escape reserved keywords to use them as identifiers
  |
7 |     () => (pub fn r#async() {})
  |                   ^^^^^^^

error: aborting due to previous error
```

**Comment by eddyb** [comment]
My guess is that the behavior *without* `-Z macro-backtrace`, to hide spans from macros in imported files, is now somehow affecting the suggestion spans, despite not doing so before.

**Comment by Centril** [comment]
I've removed the `vec![a, b, c]` ==> `[a, b, c]` suggestion since it doesn't seem to work in this new scheme.

**Comment by eddyb** [comment]
@Centril That just sounds like it just needs a reimplementation.

**Comment by Centril** [comment]
@eddyb Perhaps, but I don't know how to do it, and the existing implementation is broken in either case, (e.g. `let vec!(a, b) = [0, 1];` suggests `(a, b)`), so I think we shouldn't block on a different implementation.

File: src/librustc_expand/mbe/macro_rules.rs
Comment (eddyb) [inline src/librustc_expand/mbe/macro_rules.rs]: This should still work, but it should check for `"!["`, not just the bang (preexisting bug).

Oh look at that, this is a suggestion where the diagnostic mentions `-Z macro-backtrace`.

So this is basically the `r#async` thing, I bet that this suggestion still works but only with `-Z macro-backtrace`.

**Comment by eddyb** [comment]
So @anyska and I just figured out where the problem was, and her commit (9ea4134) has the fix.

#69739 has more details, but in short the `r#async` test was working with `-Z macro-backtrace` because that would show the imported file in the main error, and that part of diagnostics did have `ensure_source_file_source_present`, so it loaded the `SourceFile`'s `external_src`.
But without `-Z macro-backtrace`, only the suggestion logic was looking at `external_src`, which was stuck in the "not yet attempted to be loaded" state, and there was no attempt to load it made.

The `vec![` suggestion was crashing `rustfix` because of a similarly missing `ensure_source_file_source_present`, but in the JSON emitter instead.

<hr/>

With that out of the way, we can probably land this? Just needs one last rebase.

**Comment by Centril** [comment]
@anyska @eddyb Thanks a bunch for figuring this out. I've rebased the branch now, with tests working out locally. This should be ready to go now I think.

**Review by eddyb — APPROVED** [review]
LGTM, I think we can r+ once @petrochenkov has another look and CI confirms tests are passing now.

File: src/test/ui/proc-macro/span-api-tests.rs
Comment (eddyb) [inline src/test/ui/proc-macro/span-api-tests.rs]: According to https://github.com/rust-lang/rust/pull/66364#issuecomment-590095860, this is likely the only thing we can do for now.

Can you open an issue about perhaps testing the case `assert_fake_source_file` was used to test (`SourceFile::is_real()` returning `false`) and leaving a FIXME comment here mentioning that issue and/or in the `auxiliary` file where `assert_fake_source_file` is defined?

Maybe it doesn't matter that much, but I wanted to come back to this before I forget again about it.

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr
Comment (eddyb) [inline src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: So it shouldn't block this PR, but I just realized that nowhere does it actually show the *file* for this suggestion (it's in the JSON AFAIK, but that's because the JSON diagnostics are much more uniform).

Do non-suggestion "child diagnostics" get to print their own file/line/column?

cc @rust-lang/wg-diagnostics

Comment (eddyb) [reply src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: Elsewhere in the diff there's this, mid-diagnostic, to "switch files":
```diff
-  ::: <::ping::ping macros>:1:1
+  ::: $DIR/auxiliary/ping.rs:5:1
```

So it is plausible that suggestions could also check if they need to show this.

**Comment by petrochenkov** [comment]
LGTM.

**Review by petrochenkov — APPROVED** [review]

File: src/test/ui/proc-macro/span-api-tests.rs
Comment (Centril) [reply src/test/ui/proc-macro/span-api-tests.rs]: Left a FIXME: https://github.com/rust-lang/rust/issues/69775.

**Comment by Dylan-DPC** [comment]
bumping this so that we can merge it after sitting for days in the queue :D 

@bors p=1001

**Comment by pietroalbini** [comment]
@bors retry

Yield to the beta release

**Comment by Mark-Simulacrum** [comment]
Tagging as relnotes (so we don't forget about it -- it may not go into the final version), though that may be a false positive. @tomaka reported some breakage here (https://twitter.com/tomaka17/status/1238204339505770499) though it may be entirely in unstable features and unobservable on stable, in which case we wouldn't need to document it in relnotes.

@eddyb tells me @petrochenkov is the person to ask on whether there's something to say in relnotes here :)

**Comment by Mark-Simulacrum** [comment]
Further discussion on Discord informed me that perhaps there is not a significant change here and the only exposed changes are to unstable APIs in proc_macro anyway, so untagging.

**Comment by petrochenkov** [comment]
Spans can affect many things, starting from lints (as this PR itself shows).
I guess the beta crater run will show the full extent of changes this PR caused.

**Comment by petrochenkov** [comment]
Actually, running a separate crater check for this PR would be reasonable.

**Pull request merged**
